flang/lib/Semantics/check-allocate.cpp

```diff
@@ -418,14 +418,18 @@ static bool HaveCompatibleTypeParameters(
   }
 }
 
-static bool HaveCompatibleLengths(
-    const DeclTypeSpec &type1, const DeclTypeSpec &type2) {
+static bool HaveCompatibleLengths(const DeclTypeSpec &type1,
+    const DeclTypeSpec &type2, const Symbol &symbol) {
   if (type1.category() == DeclTypeSpec::Character &&
       type2.category() == DeclTypeSpec::Character) {
     auto v1{
         evaluate::ToInt64(type1.characterTypeSpec().length().GetExplicit())};
     auto v2{
         evaluate::ToInt64(type2.characterTypeSpec().length().GetExplicit())};
+    if (!type1.characterTypeSpec().length().isDeferred() && IsPointer(symbol) &&
```

Comment:
Shouldn't this condition be symmetric?  It looks like it can return true or false depending on the order of the type1 and type2 arguments.

Reply:
I think that would be true only if the check is used outside of the scope of this file. If I understand correctly, type2 will always be the type-spec parameter in this context.

Reply:
Rename the arguments, make the predicate symmetric, or at least add a comment.